cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/rolling/RackRolling.java

```diff
@@ -249,67 +315,124 @@ static Set<KafkaNode> pickBestBatchForRestart(List<Set<KafkaNode>> batches, int
      * </ol>
      *
      * @param rollClient The roll client
+     * @param contextMap The ids of the nodes in the cluster mapped to its context
      * @param nodesNeedingRestart The ids of the nodes which need to be restarted
+     * @param totalNumOfControllers The total number of controllers in this cluster
      * @param maxRestartBatchSize The maximum allowed size for a batch
      * @return The nodes corresponding to a subset of {@code nodeIdsNeedingRestart} that can safely be rolled together
      */
     private static Set<KafkaNode> nextBatch(Reconciliation reconciliation,
                                             RollClient rollClient,
-                                            Map<Integer, NodeRef> nodeMap,
-                                            Map<Integer, NodeRef> nodesNeedingRestart,
+                                            Map<Integer, Context> contextMap,
+                                            Map<Integer, NodeRoles> nodesNeedingRestart,
+                                            int totalNumOfControllers,
                                             int maxRestartBatchSize) {
         enum NodeFlavour {
-            NON_ACTIVE_PURE_CONTROLLER, // A pure KRaft controller node (not a broker) that is not the active controller
-            ACTIVE_PURE_CONTROLLER, // A pure KRaft controllers node (not a broker) that is the active controller
-            NON_ACTIVE_BROKERISH, // A KRaft or ZooKeeper node that is at least a broker (might be a
-            // KRafter combined node that is not the active controller)
-            ACTIVE_BROKERISH // A KRaft or ZooKeeper node that is a broker and also the active controller
+            NON_ACTIVE_PURE_CONTROLLER, // A pure KRaft controller node that is not the active controller
+            ACTIVE_PURE_CONTROLLER, // A pure KRaft controllers node that is the active controller
+            BROKER_AND_NOT_ACTIVE_CONTROLLER, // A node that is at least a broker and might be a
+            // controller (combined node) but that is not the active controller
+            BROKER_AND_ACTIVE_CONTROLLER // A node that is a broker and also the active controller
         }
 
-        int controllerId = rollClient.activeController();
+        Map<Integer, Long> quorumState = rollClient.quorumLastCaughtUpTimestamps();
+        int activeControllerId = rollClient.activeController();
```

Comment:
If this can be `< 0` (and the code below suggests it can) then we should update the Javadoc of the rollClient to say what that means/when it is `< 0`.